src/hotspot/share/opto/library_call.cpp

```diff
@@ -4443,6 +4443,40 @@ void LibraryCallKit::arraycopy_move_allocation_here(AllocateArrayNode* alloc, No
     Node* alloc_mem = alloc->in(TypeFunc::Memory);
     C->gvn_replace_by(callprojs.fallthrough_ioproj, alloc->in(TypeFunc::I_O));
     C->gvn_replace_by(init->proj_out(TypeFunc::Memory), alloc_mem);
+
+    // Replace InitializeNode's output_control uses' control to AllocateArrayNode's input control.
+    // CastIINode might created in GraphKit::new_array (in AllocateArrayNode::make_ideal_length) and later uses.
+    // CastIINode's input_control should still use InitializeNode's output_control after move allocation.
+    // Otherwise it breaks constraint assumption.
+    //
+    // Because alloc is tightly coupled, checking InitializeNode's control proj's output for
+    // for CastIINode. After finding CastIINode, actions include:
+    // 1. Replace CastIINode with AllocateArrayNode's length.
+    // 2. Create CastIINode again in arraycopy_move_allocation_here after control flow adjustion.
+    Node* init_control = init->proj_out(TypeFunc::Control);
```

Comment:
I would suggest rephrasing the comment to:
The CastIINode created in GraphKit::new_array (in AllocateArrayNode::make_ideal_length) must stay below the allocation (i.e. is only valid if the allocation succeeds): 1) replace CastIINode with AllocateArrayNode's length here 2) Create CastIINode again once allocation has moved (see below) at the end of this method

Reply:
Thanks @rwestrel !  Debug build check and comment are updated.

For "Could you use hash_find() instead? hash_find(prev) == hash_find(cur) maybe?"

Previous multiple CastIINodes are different as: 
- In GraphKit::new_array, CastIINode is only recorded for igvn but not transformed.
- In GraphKit::load_array_length, CastIINode is applied "_gvn.transform", which set TypeInt's _widen field as 3.
``` 
267  CastII  ===  263  226  [[]]  #int:0..max-3 !jvms: ZipPath::getFileName @ bci:65 (line 102)
272  CastII  ===  263  226  [[ 210 ]]  #int:0..max-3:www !jvms: ZipPath::getFileName @ bci:76 (line 103)
```

Updated commit applies "_gvn.transform" on CastIINode created after array allocation too.

Reply:
Release/fastdebug tier1/2/3 tested.
Fastdebug Tier1/2/3 with option "-XX:-TieredCompilation -Xbatch" also tested.